5단계-로또(수동) 추가 수정 리뷰 요청드립니다!

tlaqk229 opened the pull request:
미션완료 되었는데 추가 요청 받아주셔서 감사합니다!!
마지막 리뷰 잘 부탁드립니다~!

tlaqk229 committed:
fix(LottoNumber): 정적팩토리메소드 및 유효성검증 개선

tlaqk229 committed:
fix(Lotto): 생성자 유효성 검증 시점 변경을 통한 개선

ohtaeg commented on src/main/java/lotto/model/domain/Lotto.java:
```diff
@@ -22,11 +22,11 @@ public Lotto(List<Integer> numbers) {
 
     public Lotto(String lottoNumbersInput) {
         String[] lottoNumbers = lottoNumbersInput.split(LottoConstants.WIN_LOTTO_DELIMITER);
+        checkLottoNumberCount(lottoNumbers.length);
```
위치 변경 및 파라미터로 길이를 넘겨 체크하도록 👍

ohtaeg commented on src/main/java/lotto/model/domain/LottoNumber.java:
```diff
@@ -18,12 +18,13 @@ public class LottoNumber implements Comparable<LottoNumber> {
 
     public static LottoNumber getLottoNumberByInt(int lottoNumber) {
         if (lottoNumbers.get(lottoNumber) == null) {
-            return new LottoNumber(lottoNumber);
+            validateNumber(lottoNumber);
```
피드백 반영 좋습니다 👍

ohtaeg commented on src/main/java/lotto/model/domain/LottoNumber.java:
```diff
@@ -18,12 +18,13 @@ public class LottoNumber implements Comparable<LottoNumber> {
 
     private LottoNumber(int lottoNumber) {
-        setLottoNumber(lottoNumber);
+        validateNumber(lottoNumber);
+        this.lottoNumber = lottoNumber;
```
선 유효성 검사 후 세팅 👍

ohtaeg commented on src/main/java/lotto/model/domain/LottoNumber.java:
```diff
@@ -33,8 +34,10 @@ public static LottoNumber getLottoNumberByString(String lottoNumber) {
 
-    protected static boolean validateNumber(int checkValue) {
-        return checkValue >= LottoConstants.LOTTO_NUMBER_MIN && checkValue <= LottoConstants.LOTTO_NUMBER_MAX;
+    protected static void validateNumber(int checkValue) {
```
이 부분은 이제 개인적인 의견이라 참고만 해주셔도 좋습니다 :)

해당 메서드의 접근제어자를 protected로 선언해주신 의도가 궁금한데요!

보통 도메인 validation은 도메인 객체인 LottoNumber 스스로만 알면 되지 않을까 해서요!
누구나 다 사용해야하는 validation이라면 유틸이나 헬퍼성 클래스를 만들어도 괜찮지 않을까 합니다!
물론 현재 미션 규모에 맞게 유틸이나 헬퍼성 클래스는 필요가 없어보여 접근제어자를 최소화해도 좋을 것 같은데

윤지님은 어떻게 생각하시는지 궁금하네요 :)

접근제어자를 최소화 한다면 어떤 장점이 있을까요~?

tlaqk229 replied:
아래 코멘트 주신내용 대로 테스트를 위해 접근제어자를 protected로 선언했었습니다 😞  태경님 코멘트를 읽고 생각해보니 포비님이 라이브강의 때 'private 메소드의 테스트를 해야하는 경우 꼭 해당 메소드를 호출해서 테스트해야만 하는 것은 아니다. 해당 메소드를 호출하는 다른 메소드, 생성자를 테스트하는 것도 하나의 방법이다. 그리고 테스트를 위해 프로덕션 코드를 수정해야한다면 해당 메소드가 적절한 위치에 있지 않은 것일 수 있다.'라고 얘기했던 부분이 생각났습니다. 해당 유효성 검증 메소드는 생성자에서 활용되고 있어서 생성자를 호출해서 숫자 범위 유효성검증 테스트를 해보도록 수정해봤습니다!

ohtaeg commented on src/test/java/lotto/model/domain/LottoNumberTest.java:
```diff
@@ -17,11 +17,17 @@ public class LottoNumberTest {
+        LottoNumber.validateNumber(LottoConstants.LOTTO_NUMBER_MAX);
+        assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> {
+            LottoNumber.validateNumber(LottoConstants.LOTTO_NUMBER_MAX+1);
+        });
```
위 피드백과 같이 고민해보시면 좋을 것 같아요 :)

유효성 검증을 위한 테스트 좋습니다 👍 

다만 테스트를 위해 접근 제어자 변경을 하신거라면 배보다 배꼽이 커진 현상 같아서요!
개인적으로 코드가 존재하기에 테스트 코드가 존재하는거라 
테스트 코드가 존재해서 프로덕션 코드가 변경되면 안된다고 생각해서요 :)

만약 위 피드백이 반영이 된다면 해당 메서드는 private 접근제어자를 가지게 될 텐데
private 메서드는 어떻게 테스트할 수 있을지 고민해보시면 좋을 것 같아요!

ohtaeg replied:
추가로 하나의 시나리오를 검증함에 있어서 여러 검증문이 있을 경우 assertAll()을 활용해보시는 것을 추천드려요!
여러 검증문이 있을 때 검증이 실패했을 때와 
assertAll을 적용한 상태에서 검증이 실패했을 경우의 차이를 비교해보시면 좋을 것 같아요 :)

tlaqk229 replied:
assertAll()을 활용하니 여러 검증 중 실패하는 것이 있어도 테스트를 종료하지 않고 작성한 검증문을 모두 실행하는 차이점을 갖고있더라구요! 처음 테스트 코드 작성할 때 하나씩 안되는 걸 해결해가면서 전체가 통과하도록 작성했었는데, assertAll()은 중간에 테스트가 실패하더라도 중단되지 않으니 한번에 어떤 검증들에 문제가 있고 어떤 검증들은 통과했는지를 알기 좋더라구요! 유용한 팁 감사합니다!! 😃

ohtaeg reviewed (changes requested):
미션이 끝났음에도 더 피드백을 원하시는 열정이 대단하신 것 같아요 👍 👍 
다른 미션 진행을 위해 마지막으로 한번만 더 드릴려 했는데 
리뷰를 드리다보니 Request changes를 하게되었네요 😅 
고민해보면 좋을 피드백들을 드렸는데 한번 고민해보신 후 의견을 공유해주시거나
반영해주시면 좋을 것 같아요!

DM으로 다시 재요청 주시면 감사하겠습니다!
다른 미션을 진행하시면서 해당 PR은 우선순위를 낮춰 천천히 반영해주셔도 좋습니다!
화이팅이에요 🔥

tlaqk229 committed:
fix(LottoNumber): 접근제어자 변경 및 테스트코드 수정

- 테스트를 위해 접근제어자를 protected로 선언했던 부분 private으로 수정
- 접근제어자 private으로 수정함에 따라 테스트코드 수정
- 여러 검증문 assertAll로 테스트하도록 수정

ohtaeg commented on src/main/java/lotto/model/domain/LottoNumber.java:
```diff
@@ -34,7 +34,7 @@ public static LottoNumber getLottoNumberByString(String lottoNumber) {
     }
 
-    protected static void validateNumber(int checkValue) {
+    private static void validateNumber(int checkValue) {
```
접근제어 최소화 👍

ohtaeg commented on src/test/java/lotto/model/domain/LottoNumberTest.java:
```diff
@@ -13,21 +14,23 @@ public class LottoNumberTest {
-        assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> {
-            LottoNumber.validateNumber(LottoConstants.LOTTO_NUMBER_MAX+1);
-        });
+        assertAll(
```
assertAll 활용 👍

ohtaeg commented on src/test/java/lotto/model/domain/LottoNumberTest.java:
```diff
@@ -13,21 +14,23 @@ public class LottoNumberTest {
-        assertThat(LottoNumber.getLottoNumberByInt(5) == number).isTrue();
+        assertAll(
+                () -> assertThat(LottoNumber.getLottoNumberByInt(5)).isEqualTo(number),
+                () -> assertThat(LottoNumber.getLottoNumberByInt(5) == number).isTrue()
```
동일성 비교를 잘해주셨네요 👍  isSameAs()도 활용해보면 좋을 것 같아요!
추가로 isEqualTo() 메서드와 어떤 차이가 있는지 내부를 보시거나 검색하여 차이점을 알아보시는 것도 좋을 것 같아요 😄 
```
() -> assertThat(LottoNumber.getLottoNumberByInt(5)).isSameAs(number)
```

ohtaeg commented on src/test/java/lotto/model/domain/LottoNumberTest.java:
```diff
@@ -36,7 +39,9 @@ public class LottoNumberTest {
-        assertThat(number1.equals(number3)).isFalse();
+        assertAll(
+                () -> assertThat(number1.equals(number2)).isTrue(),
+                () -> assertThat(number1.equals(number3)).isFalse()
```
이 부분도 테스트 메서드를 활용해보면 좋을 것 같아요 😄 
```
() -> assertThat(number1).isNotEqualTo(number3)
```

ohtaeg reviewed (approved):
마지막까지 잘해주셨어요 💯 
소소한 코멘트를 드렸는데 확인해주시면 좋을 것 같아요 :)
그 외에 더 드릴 부분이 있는지 살펴보았는데 크게 드릴 코멘트가 없네요 👍 
고생 많으셨습니다 윤지님!! 🥇
